File: 23. Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& l) {
        if (l.size() == 0) {
            return nullptr;
        }
        while (l.size() > 1) {
            ListNode* a = l.back();
            l.pop_back();
            ListNode* b = l.back();
            l.pop_back();
            l.push_back(mergeTwo(a, b));   
        }
        return l[0];
    }
    ListNode* mergeTwo(ListNode* a, ListNode* b) {
        ListNode* ret = new ListNode();
        ListNode* cur = ret;
        while (a || b) {
            ListNode* next = a;
            if (!next || (b && b->val < a->val)) {
                next = b;
                b = b->next;
                next->next = nullptr;
                cur->next = next;
                cur = next;
            } else {
                a = a->next;
                next->next = nullptr;
                cur->next = next;
                cur = next;
            }
        }
        return ret->next;
    }
};

File: 23. Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None
        n = len(lists)
        ret = ListNode(0)
        ptr = ret
        flag = True
        while flag:
            cur = 0
            flag = False
            for i in range(n):
                if (lists[cur] == None and lists[i] is not None) or (lists[i] is not None and lists[i].val <= lists[cur].val):
                    flag = True
                    cur = i
            if lists[cur] == None:
                break
            tmp = lists[cur]
            lists[cur] = lists[cur].next
            tmp.next = None
            ptr.next = tmp
            ptr = ptr.next
        return ret.next
